beacon_node/beacon_chain/src/light_client_server_cache.rs

```diff
@@ -152,6 +168,75 @@ impl<T: BeaconChainTypes> LightClientServerCache<T> {
         Ok(())
     }
 
+    // - Is finalized
+    // - Has the most bits
+    // - Signed header at the oldest slot
+    fn persist_best_light_client_update(
+        &self,
+        store: &BeaconStore<T>,
+        block_root: &Hash256,
+        attested_block: SignedBeaconBlock<<T as BeaconChainTypes>::EthSpec>,
+        finalized_block: SignedBeaconBlock<<T as BeaconChainTypes>::EthSpec>,
+        chain_spec: &ChainSpec,
+    ) -> Result<(), BeaconChainError> {
+        let (block, _) = store
+            .get_full_block(block_root)?
+            .ok_or(BeaconChainError::DBInconsistent(format!(
+                "Block not available {:?}",
+                block_root
+            )))?
+            .deconstruct();
+
+        let beacon_state = store
+            .get_state(&block.state_root(), Some(block.slot()))?
+            .ok_or(BeaconChainError::DBInconsistent(format!(
+                "Beacon state not available {:?}",
+                attested_block.state_root()
+            )))?;
+
+        let mut attested_state = store
+            .get_state(&attested_block.state_root(), Some(attested_block.slot()))?
+            .ok_or(BeaconChainError::DBInconsistent(format!(
+                "Attested state not available {:?}",
+                attested_block.state_root()
+            )))?;
```

Comment:
This not acceptable w.r.t performance. You should extend the use of cached_parts such that you never have to get states from the store in the happy path. The function to generate updates in `LightClientUpdate::new` should not be used.

See this document for an implementation suggestion https://hackmd.io/@dapplion/BkxPbc6eQT